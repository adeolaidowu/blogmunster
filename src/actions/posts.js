import database from "../firebase/firebase";

//ADD_POST
export const addPost = (post) => ({
  type: "ADD_POST",
  post,
});

export const startAddPost = (postData = {}) => {
  return (dispatch) => {
    const {
      title = "",
      content = "",
      imageLink = "",
      createdAt = 0,
    } = postData;
    const post = { title, content, imageLink, createdAt };
    return database
      .ref("posts")
      .push(post)
      .then((ref) => {
        dispatch(
          addPost({
            id: ref.key,
            ...post,
          })
        );
      });
  };
};

//EDIT_POST
export const editPost = (id, updates) => ({
  type: "EDIT_POST",
  id,
  updates,
});

export const startEditPost = (id, updates) => {
  return (dispatch) => {
    dispatch(editPost(id, updates));
    database.ref(`posts/${id}`).update(updates);
  };
};

//DELETE_POST
export const deletePost = ({ id } = {}) => ({
  type: "DELETE_POST",
  id,
});

export const startDeletePost = ({ id } = {}) => {
  return (dispatch) => {
    database
      .ref(`posts/${id}`)
      .remove()
      .then(() => {
        dispatch(deletePost({ id }));
      });
  };
};

//SET_POSTS -- This action fetches data from db to display when page is visited
export const setPosts = (posts) => ({
  type: "SET_POSTS",
  posts,
});

export const startSetPosts = () => {
  return (dispatch) => {
    return database
      .ref("posts")
      .once("value")
      .then((snapshot) => {
        const posts = [];
        snapshot.forEach((childSnapshot) => {
          posts.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setPosts(posts));
      });
  };
};
