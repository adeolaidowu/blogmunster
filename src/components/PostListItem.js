import React from "react";
import moment from "moment";
import { history } from "../routers/AppRouter";
import { Link } from "react-router-dom";

const PostListItem = ({ id, title, content, imageLink, createdAt }) => {
  let path = "edit";
  let actionText = "Edit post";
  if (history.location.pathname === "/") {
    path = "post";
    actionText = "Read more";
  }
  return (
    <div className="home-post-item text-center">
      <Link className="home-post-item__title" to={`/${path}/${id}`}>
        <h2>{title}</h2>
      </Link>
      {imageLink && (
        <img className="home-post-item__img" src={imageLink} alt="blog-img" />
      )}
      <p>
        {content.slice(0, 100)}...{" "}
        <Link to={`/${path}/${id}`}>{actionText}</Link>
      </p>
      <small>{moment(createdAt).format("Do MMMM, YYYY")}</small>
      <hr />
    </div>
  );
};

export default PostListItem;
