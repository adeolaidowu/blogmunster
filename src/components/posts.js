import { v4 as uuidv4 } from "uuid";

//ADD_POST
export const addPost = ({
  title = "",
  content = "",
  imageLink = "",
  createdAt = 0,
} = {}) => ({
  type: "ADD_POST",
  post: {
    id: uuidv4(),
    title,
    content,
    imageLink,
    createdAt,
  },
});

//EDIT_POST
export const editPost = (id, updates) => ({
  type: "EDIT_POST",
  id,
  updates,
});

//DELETE_POST
export const deletePost = ({ id } = {}) => ({
  type: "DELETE_POST",
  id,
});
