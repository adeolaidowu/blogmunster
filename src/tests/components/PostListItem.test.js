import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import PostListItem from "../../components/PostListItem";
import posts from "../fixtures/posts";

test("should correctly render component", () => {
  const wrapper = shallow(<PostListItem {...posts[1]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
