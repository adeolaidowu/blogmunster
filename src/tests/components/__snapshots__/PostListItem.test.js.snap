// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render component 1`] = `
<div>
  <h2>
    <Link
      to="/edit/2"
    >
      anime
    </Link>
  </h2>
  <p>
    about anime
  </p>
  <small>
    -259200000
  </small>
</div>
`;
